python/adelphi/tests/integration/__init__.py

```diff
@@ -0,0 +1,161 @@
+                try:
+                    session.execute(realStmt)
+                except:
+                    log.info("Exception executing statement: {}".format(realStmt), exc_info=sys.exc_info()[0])
+
+
+    def runTestForVersion(self, version=None):
+        log.info("Testing Cassandra version {}".format(version))
+
+        client = docker.from_env()
+        container = client.containers.run(name="adelphi", remove=True, detach=True, ports={9042:9042}, image="cassandra:{}".format(version))
+
+        (cluster,session) = (None,None)
+        try:
+            (cluster,session) = self.connectToLocalCassandra()
+            self.createSchema(session)
+            self.runTestWithSchema(version)
+        finally:
+            if cluster:
+                cluster.shutdown()
+
+            if "KEEP_CONTAINER" in os.environ:
+                log.info("KEEP_CONTAINER env var set, preserving Cassandra container 'adelphi'")
+            else:
+                container.stop()
+
+            self.cleanUpVersion(version)
+
+
+    def testVersions(self):
+        versions = CASSANDRA_VERSIONS
+        if "CASSANDRA_VERSIONS" in os.environ:
+            versions = [s.strip() for s in os.environ["CASSANDRA_VERSIONS"].split(',')]
+
+        log.info("Testing the following Cassandra versions: {}".format(versions))
+
+        for version in versions:
+            self.runTestForVersion(version)
+
+
+TempDirs = namedtuple('TempDirs', 'basePath, outputPath, logPath')
+
+class AdelphiExportMixin:
+
+    def stdoutPath(self, version=None):
+        return os.path.join(self.dirs.outputPath, "{}-stdout.out".format(version))
+
+
+    def stderrPath(self, version=None):
+        return os.path.join(self.dirs.outputPath, "{}-stderr.log".format(version))
+
+
+    def outputDirPath(self, version=None):
+        return os.path.join(self.dirs.outputPath, version)
+
+
+    def makeTempDirs(self):
+        basePath = tempfile.mkdtemp()
+        outputPath = os.path.join(basePath, "output")
+        os.mkdir(outputPath)
+        logPath = os.path.join(basePath, "logs")
+        os.mkdir(logPath)
+        self.dirs = TempDirs(basePath, outputPath, logPath)
+
+
+    def runAdelphi(self, version=None):
+        log.info("Running Adelphi")
+        exportCmd = self.getExportCommand()
+        stdoutPath = self.stdoutPath(version)
+        stderrPath = self.stderrPath(version)
+        subprocess.run("adelphi {} > {} 2>> {}".format(exportCmd, stdoutPath, stderrPath), shell=True)
+        outputDirPath = self.outputDirPath(version)
+        os.mkdir(outputDirPath)
+        subprocess.run("adelphi --output-dir={} {} 2>> {}".format(outputDirPath, exportCmd, stderrPath), shell=True)
+        log.info("Adelphi completed")
+
+
+    def compareSchemas(self, version=None):
+        referencePath = self.getReferenceSchemaPath(version)
+        stdoutPath = self.stdoutPath(version)
+        rv1 = subprocess.run("diff -Z {} {}".format(stdoutPath, referencePath), shell=True)
```

Comment:
the `-Z` flag added in https://github.com/datastax/adelphi/commit/bd7c65294d5b5af35fb7aabadcc8ef465fdaf48e doesn't work on OSX, it's GNU specific.
These seem to be common options that we can use instead:

```
-b  --ignore-space-change
              Ignore changes in the amount of white space.

       -w  --ignore-all-space
              Ignore all white space.
```

At some point, we should probably move the diff'ing to Python too, maybe we could use something like this: https://docs.python.org/3/library/difflib.html

Reply:
It's a good point... I'm prolly gonna wind up just taking the -Z out entirely.  I added it trying to address a mismatch I was seeing but I'm pretty sure I misread the mismatch; it was actually caused by blank _lines_ while -Z only ignores excess whitespace after an otherwise matching line.  I think -B is what I actually wanted and that's supported on OS X diff.

But none of that really matters because I completely agree with your other comment; the longer term goal is to move off of a reliance on shell commands.  I did some experimenting with difflib while working on this PR and ultimately decided against using it.  The format is somewhat non-standard (it looks vaguely diff-like but isn't exactly the same) and more importantly it only works when applied to two strings... meaning I'd have to read the entire contents of the file into memory in order to do a diff.  I... don't like that so much, so I was gonna look for a streaming solution... but that can wait.

So yeah, very much +1 to the idea of this being a short-term solution.

Reply:
Update: the diff dependency has been removed and comparisons on that front are made entirely in Python now